pkgs/development/interpreters/python/mk-python-derivation.nix

```diff
@@ -104,6 +104,49 @@ else
 let
   inherit (python) stdenv;
 
+  validatePythonMatches = attrName: let
+    isPythonModule = drv:
+      # all pythonModules have the pythonModule attribute
+      (drv ? "pythonModule")
+      # Some pythonModules are turned in to a pythonApplication by setting the field to false
+      && (!builtins.isBool drv.pythonModule);
+    isMismatchedPython = drv: drv.pythonModule != python;
+
+    optionalLocation = let
+        pos = builtins.unsafeGetAttrPos (if attrs ? "pname" then "pname" else "name") attrs;
+      in if pos == null then "" else " in ${pos.file}:${toString pos.line}:${toString pos.column}";
+
+    leftPadName = name: against: let
+        len = lib.max (lib.stringLength name) (lib.stringLength against);
+      in lib.strings.fixedWidthString len " " name;
+
+    drvName = drv: drv.pname or drv.name;
+    warnMismatch = drv: lib.warn ''
+      Warning: Python version mismatch in '${name}':
+
+      The Python derivation '${name}' depends on a Python derivation
+      named '${drvName drv}', but the two derivations use different versions
+      of Python:
+
+          ${leftPadName name (drvName drv)} uses ${python}
+          ${leftPadName (drvName drv) name} uses ${toString drv.pythonModule}
+
+      Possible solutions:
```

Comment:
Maybe there should be additional explanation that would make it easier to determine which route should be attempted first.

e.g. "If ${drvName drv} provides an executable used during the build, move ${drvName drv}...",
"If ${drvName drv} provides a library used during build, change ${name}'s..."

I'd also make all drv names (including `${name}`) use single quotes; would require some slight rewording to remove the possessive `${name}'s`.

Reply:
What do you think about this?

```
trace: warning: Python version mismatch in 'hi-bogus':

The Python derivation 'hi-bogus' depends on a Python derivation
named 'flake8', but the two derivations use different versions
of Python:

    'hi-bogus' uses /nix/store/d44wd6n98f93hjr6q1d1phhh1hw7a17d-python3-3.8.8
      'flake8' uses /nix/store/pp0fp5zds7wqhyx6zqnnxl2aqzf4a51r-python-2.7.18

Possible solutions for /home/grahamc/projects/github.com/NixOS/nixpkgs/test.nix:5:3:

  * If this is a Python library you depend on, change the reference to 'flake8'
    in the propagatedBuildInputs of 'hi-bogus' to use a 'flake8' built from the same
    version of Python.

  * If 'flake8' provides executables, move the reference to 'flake8' in
    'hi-bogus' from propagatedBuildInputs to nativeBuildInputs.
```

Reply:
Slight wording change in the suggestions:

```
trace: warning: Python version mismatch in 'hi-bogus':

The Python derivation 'hi-bogus' depends on a Python derivation
named 'flake8', but the two derivations use different versions
of Python:

    'hi-bogus' uses /nix/store/d44wd6n98f93hjr6q1d1phhh1hw7a17d-python3-3.8.8
      'flake8' uses /nix/store/pp0fp5zds7wqhyx6zqnnxl2aqzf4a51r-python-2.7.18

Possible solutions:

  * If 'flake8' is a Python library, change its reference in the propagatedBuildInputs
    of 'hi-bogus' to use a 'flake8' built from the same version of Python.

  * If 'flake8' provides executables used while building the package, move the reference
    to 'flake8' in 'hi-bogus' from propagatedBuildInputs to nativeBuildInputs.

at /home/grahamc/projects/github.com/NixOS/nixpkgs/test.nix:5:3
```